Sources/Plasma/NucleusLib/pnNetProtocol/Private/pnNpCommon.cpp

```diff
@@ -297,605 +294,289 @@ void NetGameRank::CopyFrom(const NetGameRank & fromRank) {
-}
-
-//============================================================================
-void NetVaultNode::SetBlob_2 (const uint8_t v[], uint32_t len) {
-    IVaultNodeSetBlob(kBlob_2, &blob_2, &blob_2Length, v, len);
-}
-
 
-/*****************************************************************************
-*
-*   CSrvPackBuffer
-*
-***/
-
-//============================================================================
-CSrvPackBuffer::CSrvPackBuffer (unsigned bytes) {
-    m_data = (uint8_t *)malloc(bytes);
-    m_pos  = m_data;
-    m_end  = m_pos + bytes;
-}
-
-//============================================================================
-void * CSrvPackBuffer::Alloc (unsigned bytes) {
-    ASSERT((signed) bytes >= 0);
-    ASSERT(m_pos + bytes <= m_end);
-
-    uint8_t * pos = m_pos;
-    m_pos += bytes;
-    return pos;
-}
-
-//============================================================================
-void CSrvPackBuffer::AddData (const void * ptr, unsigned bytes) {
-    memcpy(Alloc(bytes), ptr, bytes);
-}
-
-//============================================================================
-void CSrvPackBuffer::AddString (const wchar_t str[]) {
-    AddData(str, StrBytes(str));
-}
-
-//============================================================================
-void CSrvPackBuffer::AddDWordArray (const uint32_t * arr, unsigned count) {
-    // Don't let large counts cause pointer wrap
-    count &= 0x00ffffff;
-    AddData(arr, count * sizeof(arr[0]));
-}
-
-//============================================================================
-unsigned CSrvPackBuffer::Size () {
-    return m_pos - m_data;
-}
-
-
-/*****************************************************************************
-*
-*   CSrvUnpackBuffer
-*
-***/
-
-//============================================================================
-CSrvUnpackBuffer::CSrvUnpackBuffer (const void * buffer, unsigned count) {
-    m_pos = (const uint8_t *) buffer;
-    m_end = m_pos + count;
+    // We should never get here, but this silences a warning.
+    return true;
 }
 
 //============================================================================
-const void * CSrvUnpackBuffer::GetData (unsigned bytes) {
-    for (;;) {
-        const uint8_t * result = m_pos;
-        m_pos += bytes;
-        
-        if (m_pos < result)
-            break;
-        if (m_pos > m_end)
-            break;
-            
-        return result;
+template<typename T>
+static void IRead(const uint8_t*& buf, T& dest)
+{
+    const T* ptr = reinterpret_cast<const T*>(buf);
+    dest = *ptr;
+    buf += sizeof(T);
+}
+
+template<>
+static void IRead<plString>(const uint8_t*& buf, plString& dest)
+{
+    uint32_t size = *(reinterpret_cast<const uint32_t*>(buf));
+    buf += sizeof(uint32_t);
+
+    plStringBuffer<uint16_t> str;
+    uint16_t* theStrBuffer = str.CreateWritableBuffer(size / sizeof(uint16_t));
+    memcpy(theStrBuffer, buf, size);
+    dest = plString::FromUtf16(str);
```

Comment:
NOTE:  `CreateWritableBuffer` does NOT automagically nul-terminate the string.  You should either nul-terminate the input, or explicitly specify the size of the input in the `FromUtf16` call.